File: task3_8/main.py
import httpx
from fastapi import FastAPI, HTTPException

app = FastAPI(
    title="API Gateway Service",
    description="Интеграционный сервис для связи с Go-бэкендом",
    version="1.0.0"
)

GO_SERVICE_URL = "http://localhost:8080"

async def make_request(method, path, json=None):
    async with httpx.AsyncClient() as client:
        try:
            url = f"{GO_SERVICE_URL}{path}"
            response = await client.request(method, url, json=json)
            response.raise_for_status()
            return response.json()
        except httpx.HTTPStatusError as e:
            raise HTTPException(
                status_code=e.response.status_code,
                detail=f"Go service error: {e.response.text}"
            )
        except httpx.RequestError:
            raise HTTPException(
                status_code=503,
                detail="Go service unavailable"
            )

@app.get("/fetch-status", tags=["Go Proxy"])
async def get_go_status():
    data = await make_request("GET", "/status")
    return {"source": "go-service", "data": data}

@app.get("/fetch-info", tags=["Go Proxy"])
async def get_go_info():
    return await make_request("GET", "/info")

@app.post("/send-data", tags=["Go Proxy"])
async def send_to_go(payload: dict):
    return await make_request("POST", "/data", json=payload)

File: task3_8/main.go
package main

import (
	"fmt"
	"time"

	"github.com/gin-gonic/gin"
	swaggerFiles "github.com/swaggo/files"
	ginSwagger "github.com/swaggo/gin-swagger"
	_ "go_service/docs"
)

// @title           Go Gin API
// @version         1.0
// @description     Сервер на Gin с поддержкой Swagger (Задание 8).
// @host            localhost:8080
// @BasePath        /

func LoggerMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		start := time.Now()
		c.Next()
		latency := time.Since(start)
		fmt.Printf("[LOG] %s %s %d %v\n", c.Request.Method, c.Request.URL.Path, c.Writer.Status(), latency)
	}
}

// @Summary      Проверка статуса
// @Tags         System
// @Produce      json
// @Success      200  {object}  map[string]string
// @Router       /status [get]
func StatusHandler(c *gin.Context) {
	c.JSON(200, gin.H{"status": "ok"})
}

// @Summary      Информация о сервисе
// @Tags         System
// @Produce      json
// @Success      200  {object}  map[string]string
// @Router       /info [get]
func InfoHandler(c *gin.Context) {
	c.JSON(200, gin.H{"service": "go-gin", "version": "1.0"})
}

// DataHandler godoc
// @Summary      Отправка данных
// @Tags         Data
// @Accept       json
// @Produce      json
// @Param        payload  body      object  true  "JSON данные для отправки"
// @Success      200      {object}  map[string]string
// @Router       /data [post]
func DataHandler(c *gin.Context) {
    c.JSON(200, gin.H{"message": "data received"})
}

func SetupRouter() *gin.Engine {
	r := gin.New()
	r.HandleMethodNotAllowed = true
	r.Use(LoggerMiddleware())
	r.Use(gin.Recovery())

	r.GET("/swagger/*any", ginSwagger.WrapHandler(swaggerFiles.Handler))

	r.GET("/status", StatusHandler)
	r.GET("/info", InfoHandler)
	r.POST("/data", DataHandler)

	return r
}

func main() {
	r := SetupRouter()
	r.Run(":8080")
}